p2p/pqueue.go

```diff
@@ -0,0 +1,278 @@
+}
+
+func (s *pqScheduler) close() {
+	s.closer.Close()
+	<-s.done.Done()
+}
+
+func (s *pqScheduler) closed() <-chan struct{} {
+	return s.closer.Done()
+}
+
+// start starts non-blocking process that starts the priority queue scheduler.
+func (s *pqScheduler) start() {
+	go s.process()
+}
+
+// process starts a block process where we listen for Envelopes to enqueue. If
+// there is sufficient capacity, it will be enqueued into the priority queue,
+// otherwise, we attempt to dequeue enough elements from the priority queue to
+// make room for the incoming Envelope by dropping lower priority elements. If
+// there isn't sufficient capacity at lower priorities for the incoming Envelope,
+// it is dropped.
+//
+// After we attempt to enqueue the incoming Envelope, if the priority queue is
+// non-empty, we pop the top Envelope and send it on the dequeueCh.
+func (s *pqScheduler) process() {
+	defer s.done.Close()
+
+	for {
+		select {
+		case e := <-s.enqueueCh:
+			chIDStr := strconv.Itoa(int(e.channelID))
+			pqEnv := &pqEnvelope{
+				envelope:  e,
+				size:      uint(proto.Size(e.Message)),
+				priority:  s.chPriorities[e.channelID],
+				timestamp: time.Now().UTC(),
+			}
+
+			// enqueue
+
+			// Check if we have sufficient capacity to simply enqueue the incoming
+			// Envelope.
+			if s.size+pqEnv.size <= s.capacity {
+				// enqueue the incoming Envelope
+				s.push(pqEnv)
+			} else {
+				// There is not sufficient capacity to simply enqueue the incoming
+				// Envelope. So we have to attempt to make room for it by dropping lower
+				// priority Envelopes or drop the incoming Envelope otherwise.
+				var total uint
+
+				for i := 0; i < len(s.chDescs) && uint(s.chDescs[i].Priority) < pqEnv.priority && total < pqEnv.size; i++ {
+					total += s.sizes[uint(s.chDescs[i].Priority)]
+				}
+
+				if total >= pqEnv.size {
+					// There is room for the incoming Envelope, so we drop as many lower
+					// priority Envelopes as we need to.
+					var (
+						canEnqueue bool
+						tmpSize    = s.size
+						i          = s.pq.Len() - 1
+					)
+
+					// Drop lower priority Envelopes until sufficient capacity exists for
+					// the incoming Envelope
+					for i >= 0 && !canEnqueue {
+						pqEnvTmp := s.pq.get(i)
+
+						if pqEnvTmp.priority < pqEnv.priority {
+							if tmpSize+pqEnv.size <= s.capacity {
+								canEnqueue = true
+							} else {
+								pqEnvTmpChIDStr := strconv.Itoa(int(pqEnvTmp.envelope.channelID))
+								s.metrics.PeerQueueDroppedMsgs.With("ch_id", pqEnvTmpChIDStr).Add(1)
+								s.logger.Debug(
+									"dropped envelope",
+									"ch_id", pqEnvTmpChIDStr,
+									"priority", pqEnvTmp.priority,
+									"msg_size", pqEnvTmp.size,
+									"capacity", s.capacity,
+								)
+
+								// dequeue/drop from the priority queue
+								heap.Remove(s.pq, pqEnvTmp.index)
+
+								// update the size tracker
+								tmpSize -= pqEnvTmp.size
+
+								// start from the end again
+								i = s.pq.Len() - 1
```

Comment:
If the queue is ordered and you've just removed the one at the end (which has the lowest priority) than can't you just decrement i and continue to remove the next one until you reach an adequate `tmpSize`

Reply:
The queue is _not_ ordered. It's a max heap implemented via a priority queue. At any given node, the only thing you can assume is that all descendants are of lower priority, but the underlying array/slice is not ordered.